pkg/sqlutil/monitor.go

```diff
@@ -116,32 +116,82 @@ func (q *queryLogger) logError(err error) {
 	q.lggr.Errorw("SQL ERROR", "err", err, "sql", q)
 }
 
-// logTiming logs about context cancellation and timing after a query returns.
-// Queries which use their full timeout log critical level. More than 50% log error, and 10% warn.
+// logTiming logs about context errors and timing.
+// To be deferred to run after a query returns.
 func (q *queryLogger) logTiming(ctx context.Context, start time.Time) {
 	elapsed := time.Since(start)
-	if ctx.Err() != nil {
-		q.lggr.Debugw("SQL CONTEXT CANCELLED", "ms", elapsed.Milliseconds(), "err", ctx.Err(), "sql", q)
+
+	if err := ctx.Err(); err != nil {
+		if errors.Is(err, context.Canceled) {
+			q.lggr.Debugw("SQL Context Canceled", "ms", elapsed.Milliseconds(), "err", context.Cause(ctx), "sql", q)
+			return
+		} else if !errors.Is(err, context.DeadlineExceeded) {
+			q.lggr.Debugw("SQL Context Error", "ms", elapsed.Milliseconds(), "err", err, "sql", q)
+			return
+		}
+		q.lggr.Debugw("SQL Deadline Exceeded", "ms", elapsed.Milliseconds(), "err", err, "sql", q)
 	}
 
+	// Success or Deadline Exceeded, so calculate how much of the timeout was used.
 	deadline, ok := ctx.Deadline()
 	if !ok {
 		return
 	}
 	timeout := deadline.Sub(start)
 
-	pct := float64(elapsed) / float64(timeout)
-	pct *= 100
+	var thresholds LogThresholds
+	thresholds.FromContextValue(ctx)
+
+	var pct float64 // percent of timeout used
+	if timeout > 0 {
+		pct = float64(elapsed) / float64(timeout)
+		pct *= 100
+	} else {
+		timeout = 0
+		pct = 100
+	}
 
 	kvs := []any{"ms", elapsed.Milliseconds(), "timeout", timeout.Milliseconds(), "percent", strconv.FormatFloat(pct, 'f', 1, 64), "sql", q}
 
 	if elapsed >= timeout {
 		q.lggr.Criticalw(slowMsg, kvs...)
-	} else if errThreshold := timeout / 5; errThreshold > 0 && elapsed > errThreshold {
+	} else if errThreshold := thresholds.Error(timeout); errThreshold > 0 && elapsed > errThreshold {
 		q.lggr.Errorw(slowMsg, kvs...)
-	} else if warnThreshold := timeout / 10; warnThreshold > 0 && elapsed > warnThreshold {
+	} else if warnThreshold := thresholds.Warn(timeout); warnThreshold > 0 && elapsed > warnThreshold {
 		q.lggr.Warnw(slowMsg, kvs...)
 	}
 
 	PromSQLQueryTime.Observe(pct)
 }
+
+// TODO document
+type LogThresholds struct {
+	Warn  func(time.Duration) time.Duration
+	Error func(time.Duration) time.Duration
+}
+
+type ctxKeyLogThresholds struct{}
+
+func (t *LogThresholds) ContextWithValue(ctx context.Context) context.Context {
+	if t == nil {
+		return ctx
+	}
+	return context.WithValue(ctx, ctxKeyLogThresholds{}, *t)
```

Comment:
Cool! Glad to see us using values in contexts finally, I've wondered why we don't use them more.

Although I hope this won't interfere with any future use of them. If some component needs to have a value attached to a context, but also needs it to have a deadline with custom thresholds, is there a clean way for those to co-exist?

Reply:
The context key used here is an unexported type, so nothing outside of this package can use the same key and conflict with it.